### PR TITLE
HPCC-16649 Restrict moving projects over distributes if row size increases

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:

Run full compiler regression tests.  
Some changes: aslow2, aslow4, childdistribute, thorchildcall,tina2,jake41

### CODE REVIEW

**Comment by richardkchapman** [comment]
Typo (restict) in commit message

**Comment by ghalliday** [comment]
@shamser please review.

**Comment by ghalliday** [comment]
github seems to be not reflecting changes immediately.  commit message has been changed, and the #if 0 cleaned up

File: ecl/hql/hqlattr.cpp [hunk 1]
```diff
@@ -2046,37 +2046,56 @@ static unsigned estimateRowSize(IHqlExpression * record)
...
+    if (newStats.datasetFields > oldStats.datasetFields)
+        return true;
+    if (newStats.dictionaryFields > oldStats.dictionaryFields)
+        return true;
+
     return false;
 }
 
-bool increasesRowSize(IHqlExpression * expr)
+bool reducesRowSize(IHqlExpression * expr)
```
Comment (shamser) [inline ecl/hql/hqlattr.cpp]: The set of tests used to check if the row size is likely to be similiar is a different set to increasesRowSize.  Should the same set of tests be used?  If not may be worth documenting why increasesRowSize isn't inverse to decreasesRowSize

**Review by shamser — COMMENTED** [review]
@ghalliday A few comments in the review.

File: ecl/hql/hqlopt.cpp [hunk 2]
```diff
@@ -3325,14 +3325,14 @@ IHqlExpression * CTreeOptimizer::doCreateTransformed(IHqlExpression * transforme
             case no_subsort:
                 if (transformedCountProject)
                     break;
-                if (increasesRowSize(transformed))
+                if ((options & HOOminimizeNetworkAndMemory) && increasesRowSize(transformed))
```
Comment (ghalliday) [inline ecl/hql/hqlopt.cpp]: Moving a project over a sort is often a good idea if that project can be combined with a disk read.

If the rows are definitely held in memory then it doesn't really matter how big the rows that are sorted are - and there is more benefit from combining the project with other operations.  hence the new flag.

Comment (ghalliday) [inline ecl/hql/hqlopt.cpp]: If the new row size is unknown it should default to assuming the size is increased.  I didn't come up with a good way of naming the function.

File: ecl/hql/hqlattr.cpp (hunk 1 above)
Comment (ghalliday) [inline ecl/hql/hqlattr.cpp]: I'll think about how they are related.  You are right they should be similar/the same.

**Comment by shamser** [comment]
@ghalliday There seems to be some merge conflicts now.

**Comment by ghalliday** [comment]
@shamser I've improved the logic, and ensured it is used for increasesRowSize and reducesRowSize().  Latest changes are a separate commit, but I will squash once you're happy.

I also added a HPCC-17606 to cover improving it further.

**Comment by shamser** [comment]
@ghalliday Looks good to me.

**Pull request merged**